### PR TITLE
Refactore baseline fs2 client bench

### PR DESCRIPTION
**Pull request by barambani**
This give another bit of info
```
Benchmark                                               Mode  Cnt      Score     Error  Units
ParallelLoadLaserdiscCatsBench.parallelLoad1           thrpt  200   5291.802 ±  14.365  ops/s
ParallelLoadLaserdiscCatsBench.parallelLoad2           thrpt  200  13374.111 ± 117.475  ops/s
ParallelLoadLaserdiscCatsRespBench.parallelLoad1       thrpt  200   5951.609 ±  56.826  ops/s
ParallelLoadLaserdiscCatsRespBench.parallelLoad2       thrpt  200  10264.231 ±  61.130  ops/s
ParallelLoadLaserdiscCatsBitVectorBench.parallelLoad1  thrpt  200   7296.143 ±  33.001  ops/s
ParallelLoadLaserdiscCatsBitVectorBench.parallelLoad2  thrpt  200  11417.866 ± 120.308  ops/s
ParallelLoadLaserdiscCatsByteBench.parallelLoad1       thrpt  200  48985.944 ± 478.882  ops/s
ParallelLoadLaserdiscCatsByteBench.parallelLoad2       thrpt  200  52861.300 ± 143.582  ops/s
ParallelLoadLaserdiscCatsBaselineBench.parallelLoad1   thrpt  200  49127.852 ± 349.831  ops/s
ParallelLoadLaserdiscCatsBaselineBench.parallelLoad2   thrpt  200  53043.959 ± 145.580  ops/s
```
in particular what's interesting is the difference 
```
ParallelLoadLaserdiscCatsByteBench.parallelLoad1       thrpt  200  48985.944 ± 478.882  ops/s
ParallelLoadLaserdiscCatsByteBench.parallelLoad2       thrpt  200  52861.300 ± 143.582  ops/s
ParallelLoadLaserdiscCatsBaselineBench.parallelLoad1   thrpt  200  49127.852 ± 349.831  ops/s
ParallelLoadLaserdiscCatsBaselineBench.parallelLoad2   thrpt  200  53043.959 ± 145.580  ops/s
```
where baseline goes through `Pipe[F, Byte, Byte]` and byte goes through `Pipe[F, Byte, BitVector]`. The latter show that the framing and the scodec processing involved are not that costly.

### CODE REVIEW

File: benchmarks/fs2/src/main/scala/laserdisc/fs2/parallel/baseline/BitVectorChannelCodec.scala [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+import _root_.fs2.{Chunk, Pipe, Pull, Stream}
+import cats.MonadError
+import laserdisc.protocol._
+import scodec.bits.BitVector
+
+private[fs2] object BitVectorChannelCodec {
+  def send[F[_]: MonadError[*[_], Throwable]](socketChannel: Pipe[F, Byte, Unit]): Pipe[F, BitVector, Unit] =
+    _.flatMap(bits => Stream.chunk(Chunk.bytes(bits.toByteArray)))
+      .through(socketChannel)
```
Comment (sirocchj) [inline benchmarks/fs2/src/main/scala/laserdisc/fs2/parallel/baseline/BitVectorChannelCodec.scala:14]: so something in here is fishy...

Comment (barambani) [reply benchmarks/fs2/src/main/scala/laserdisc/fs2/parallel/baseline/BitVectorChannelCodec.scala:14]: my target now is `toByteArray`. We'll see.

**Comment by barambani** [comment]
I'll do the renaming in this PR.

**Comment by sirocchj** [comment]
Been doing some playing around myself... have promising results for `parallelLoad2` but not huge improvements for `parallelLoad1` case.. it involves getting chunks and then evalMap them building the Array[Byte] with all chunks’ BitVector

**Comment by barambani** [comment]
the case 1 is kind of asymptotic. The common use case should be 2. The size of the payloads in 1 is more to stress the client I guess. Anyway would be ideal to serve decently that one as well.

**Comment by sirocchj** [comment]
Yeah, so I see the following (based off master not this PR):
```
[info] Benchmark                                              Mode  Cnt      Score      Error  Units
[info] ParallelLoadLaserdiscCatsBaselineBench.parallelLoad1    thrpt   20  64619.136 ±  588.890  ops/s
[info] ParallelLoadLaserdiscCatsBaselineBench.parallelLoad2    thrpt   20  71287.098 ± 1136.488  ops/s
[info] ParallelLoadLaserdiscCatsBitVectorBench.parallelLoad1   thrpt   20   9759.063 ±  66.975  ops/s
[info] ParallelLoadLaserdiscCatsBitVectorBench.parallelLoad2   thrpt   20  11754.853 ± 143.012  ops/s
[info] ParallelLoadLaserdiscCatsBitVectorBench2.parallelLoad1  thrpt   20  13665.831 ± 145.620  ops/s
[info] ParallelLoadLaserdiscCatsBitVectorBench2.parallelLoad2  thrpt   20  47558.596 ± 871.871  ops/s
[info] ParallelLoadLaserdiscCatsByteBench.parallelLoad1        thrpt   20  67636.546 ± 384.506  ops/s
[info] ParallelLoadLaserdiscCatsByteBench.parallelLoad2        thrpt   20  72386.290 ± 959.776  ops/s
```
That looks like a 40% speedup in the asymptotic case and a 400% speedup in the common one over version based on what we have in master.
This is the change:
```diff
% diff baseline/RedisBitVectorChannel.scala baseline/RedisBitVectorChannel2.scala
18c18
< object RedisBitVectorChannel {
---
> object RedisBitVectorChannel2 {
30c30,36
<         val send    = stream.through(impl.send(socket.writes(writeTimeout)))
---
>         val send    = stream.chunks.evalMap { chunk =>
>           val ab = new scala.collection.mutable.ArrayBuffer[Byte]()
>           chunk.foreach { bits =>
>             ab ++= bits.toByteArray
>           }
>           socket.write(Chunk.bytes(ab.toArray), writeTimeout)
>         }
41,43d46
<     def send[F[_]: MonadError[*[_], Throwable]](socketChannel: Pipe[F, Byte, Unit]): Pipe[F, BitVector, Unit] =
<       _.flatMap(bits => Stream.chunk(Chunk.bytes(bits.toByteArray)))
<         .through(socketChannel)
```
I then tried concatenating the `BitVectors` first and only after `.toByteArray` them and here's the results:
```
[info] ParallelLoadLaserdiscCatsBitVectorBench3.parallelLoad1  thrpt   20  16215.854 ± 154.789  ops/s
[info] ParallelLoadLaserdiscCatsBitVectorBench3.parallelLoad2  thrpt   20  48187.637 ± 516.024  ops/s
```
Doesn't do much in the common case (additional 10%) but it appears to benefit more the asymptotic case (improving it by an additional 25%).
Here's the change:
```diff
% diff baseline/RedisBitVectorChannel.scala baseline/RedisBitVectorChannel3.scala
18c18
< object RedisBitVectorChannel {
---
> object RedisBitVectorChannel3 {
30c30,34
<         val send    = stream.through(impl.send(socket.writes(writeTimeout)))
---
>         val send    = stream.chunks.evalMap { chunk =>
>           socket.write(Chunk.bytes(chunk.foldLeft(BitVector.empty)(_ ++ _).toByteArray), writeTimeout)
>         }
41,43d44
<     def send[F[_]: MonadError[*[_], Throwable]](socketChannel: Pipe[F, Byte, Unit]): Pipe[F, BitVector, Unit] =
<       _.flatMap(bits => Stream.chunk(Chunk.bytes(bits.toByteArray)))
<         .through(socketChannel)
```
I'm sure there's more to squeeze...

**Pull request merged**